components/driver/spi_master.c

```diff
@@ -811,15 +811,32 @@ esp_err_t SPI_MASTER_ATTR spi_device_queue_trans(spi_device_handle_t handle, spi
     esp_pm_lock_acquire(host->bus_attr->pm_lock);
 #endif
     //Send to queue and invoke the ISR.
-
-    BaseType_t r = xQueueSend(handle->trans_queue, (void *)&trans_buf, ticks_to_wait);
-    if (!r) {
-        ret = ESP_ERR_TIMEOUT;
-#ifdef CONFIG_PM_ENABLE
-        //Release APB frequency lock
-        esp_pm_lock_release(host->bus_attr->pm_lock);
-#endif
-        goto clean_up;
+    if (from_isr) {
+        if (xQueueIsQueueFullFromISR(handle->trans_queue) == pdTRUE && uxQueueMessagesWaitingFromISR(handle->trans_queue) == 1) {
```

Comment:
The alternative is less desirable, that's why. Overwriting happens if the queue only contains 1 element and it's full, so if not overwriting, then the transaction can't be queued, so it would fail (we are in a ISR, we can't wait). In the usage case, that is when called from ISR, it's likely the new transaction is more desirable (since it came after the previous enqueued transaction that hasn't started yet, since it wasn't dequeued). 

As I've written above, in the queue size==1 case, the queue is only used to pass a transaction from one ISR to another, not as a real queue, so we want atomic write here, and xQueueOverwrite is the closest to the atomic write primitive.

Reply:
This checking is not safe/not necessary. Please note that this function may be called in a different core from the SPI ISR.

See
```mermaid
flowchart
subgraph spi_queue_isr
check["check uxQueueMessagesWaitingFromISR(handle->trans_queue) == 1"]
queue_override["xQueueOverwriteFromISR"]

check --> queue_override
end

subgraph spi_isr
queue_receive["QueueReceiveFromISR"]
start_transaction
queue_receive --> start_transaction
end

check -.->|then| queue_receive -.->|then| queue_override 

```

So it's either unsafe (need to check the implemtation of FreeRTOS queue, or the driver), or you should always `xQueueOverwriteFromISR` without checking anything.

Reply:
And a more important question, how does the user know the previous transaction is overwritten but not executed?

This implementation doesn't look like to fit into the current architecture - transactions can be queued then executed. But just limiting a lot of use cases just to get rid of the problem of handling errors.

Reply:
I'm not following you here. The code that does this double checking is only called when you're already in an ISR context, so there isn't any multi/thread/task *concurrent* scenario possible here. The function is now static (not visible to the user code), and the last parameter `from_isr` tells if it's called from an ISR or not. It's only when it's called from an ISR that this double check is asserted, so it can't be interrupted in the middle like in your scenario. 

This parameter `from_isr` is only true when the user call the new public function `spi_device_queue_trans_from_isr` so, IMHO, it's quite clear that they are not using the standard function but know what they are doing. The previous `spi_device_queue_trans` function will set `from_isr = false`  so such tests will not happen, and it'll behave as usual with a queue send and not an overwrite.

So, unless I'm mistaken, for 99% of the users, this PR will not change anything, it'll not create multi-threading/tasking issue, since it'll work as it previously worked. But for those 1% that need a SPI transaction to happen as fast as possible from an interrupt context, this PR allows it, and in that case, you are loosing the ability to queue multiple transaction in a full queue since FreeRTOS doesn't allow a blocking queue appending from an ISR. I can't change this (this is a FreeRTOS limitation).

I can remove the check and simply call `xQueueSendToBackFromISR` but it will force us to create a large queue (size > 1) where only one item is actually required, this is not optimal IMHO. 

I've documented this behavior `Since transactions are sent to a queue from an ISR, they can't wait if the queue is full. So this method will either overwrite the queue content if it only contains one item (recommended), or fail to push if full.`, and I think it's the most intuitive, in the case where triggering a SPI transaction should happen in an ISR, you can trigger the first transaction this way. If you need to trigger multiple transactions, you'll trigger the first transaction this way and in the SPI's transaction done's callback (called from the SPI's ISR), queue the other transactions one by one.

Said differently, if you work *the ISR way*, you simply use FreeRTOS Queue as an atomic primitive containing one element and do the queueing work yourself, but you gain a huge drop in latency (which is required for some sensors to work).